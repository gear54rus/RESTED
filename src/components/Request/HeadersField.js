import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';
import { FormGroup, FormControl, Row, Col } from 'react-bootstrap';

import Fonticon from 'components/Fonticon';
import Collapsible from 'components/Collapsible';

import HeaderNameAutosuggest from './HeaderNameAutosuggest';
import { UnstyledButton, TrashButton } from './StyledComponents';

function renderValueField({ input, placeholder }) {
  return (
    <FormControl
      type="text"
      placeholder={placeholder}
      {...input}
    />
  );
}

renderValueField.propTypes = {
  input: PropTypes.shape({}).isRequired,
  placeholder: PropTypes.string.isRequired,
};

function HeadersField({ meta, fields }) {
  return (
    <Collapsible
      title="Headers"
      id="headers"
    >
      {fields.map((field, key) => (
        <FormGroup
          key={key}
          controlId={`header.${field}`}
          validationState={meta.invalid ? 'error' : null}
        >
          <Col xs={5}>
            <Field
              name={`${field}.name`}
              component={HeaderNameAutosuggest}
              placeholder="Name"
            />
          </Col>
          <Col xs={5}>
            <Field
              name={`${field}.value`}
              component={renderValueField}
              placeholder="Value"
            />
          </Col>
          <Col xs={2}>
            <TrashButton
              id={`removeHeaderButton${key}`}
              tooltip="Remove header"
              icon="trash"
              onClick={e => {
                e.preventDefault();
                fields.remove(key);
              }}
            />
          </Col>
        </FormGroup>
      ))}

      <Row>
        <Col xs={12}>
          <UnstyledButton
            id="addHeaderButton"
            bsStyle="link"
            onClick={() => fields.push({})}
          >
            <Fonticon icon="plus" />
            Add header
          </UnstyledButton>
        </Col>
      </Row>
    </Collapsible>
  );
}

// ESLint is not smart enough to see that these are used
/* eslint-disable react/no-unused-prop-types */
HeadersField.propTypes = {
  fields: PropTypes.shape({
    map: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired,
  }).isRequired,
  meta: PropTypes.shape({
    invalid: PropTypes.bool.isRequired,
  }).isRequired,
};
/* eslint-enable react/no-unused-prop-types */

export default HeadersField;

