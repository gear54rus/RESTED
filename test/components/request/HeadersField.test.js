import React from 'react';
import renderer from 'react-test-renderer';
import { combineReducers, createStore } from 'redux';
import { FieldArray, reduxForm, reducer as form } from 'redux-form';
import { Provider, connect } from 'react-redux';
import { mount } from 'enzyme';

/* eslint-disable import/no-unresolved */
import HeadersField from 'components/Request/HeadersField';
import config from 'store/config/reducer';

const makeStore = (initial = {}) => createStore(
  combineReducers({ config, form }),
  { config: initial },
);

describe('HeadersField', () => {
  let store;
  let props;
  const headers = [
    {
      name: 'Content-Type',
      value: 'application/awesome',
    }, {
      name: 'DNT',
      value: '1',
    }, {
      name: 'Accept',
      value: 'text/html',
    },
  ];

  beforeEach(() => {
    props = {
      fields: {
        map: headers.map.bind(headers),
        remove: jest.fn(),
        push: jest.fn(),
      },
      meta: {
        invalid: false,
        valid: true,
      },
    };

    store = makeStore({
      config: {},
      testForm: {},
    });
  });

  it('should match the previous snapshot', () => {
    // Mock props
    HeadersField.defaultProps = props;

    const Decorated = reduxForm({
      form: 'testForm',
    })(HeadersField);

    const tree = renderer.create(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should render the fields with correct types', () => {
    let Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <FieldArray
        name="headers"
        component={HeadersField}
      />
    ));

    Decorated = connect(() => ({
      initialValues: {
        headers,
      },
    }))(Decorated);

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    const inputs = tree.find('input');
    expect(inputs.length).toBe(6);

    let i = 0;
    headers.forEach(header => {
      const nameField = inputs.at(i++); // eslint-disable-line no-plusplus
      const valField = inputs.at(i++); // eslint-disable-line no-plusplus

      expect(nameField.prop('placeholder')).toBe('Name');
      expect(valField.prop('placeholder')).toBe('Value');
      expect(nameField.prop('value')).toBe(header.name);
      expect(valField.prop('value')).toBe(header.value);
    });
  });

  it('should call fields.push when "Add header" is pressed', () => {
    const Decorated = reduxForm({
      form: 'testForm',
    })(() => <HeadersField {...props} />);

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    expect(props.fields.push).not.toHaveBeenCalled();

    tree.find('button#addHeaderButton').simulate('click');
    expect(props.fields.push).toHaveBeenCalledWith({});
  });

  it('should call fields.remove when "Remove header" is pressed', () => {
    const Decorated = reduxForm({
      form: 'testForm',
    })(() => <HeadersField {...props} />);

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    expect(props.fields.remove).not.toHaveBeenCalled();

    tree.find('button#removeHeaderButton0').simulate('click');
    expect(props.fields.remove).toHaveBeenCalledWith(0);
  });
});

