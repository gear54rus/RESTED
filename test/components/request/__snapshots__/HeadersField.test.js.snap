// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadersField should match the previous snapshot 1`] = `
<div>
  <button
    className="sc-htpNat iPErUs btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <h4
      className="sc-bdVaJa bcSwbK"
    >
      Headers
      <i
        className="fas fa-angle-right sc-bwzfXH dSpMtd"
        role="presentation"
      />
    </h4>
  </button>
  <div
    aria-expanded={null}
    className="collapse row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="form-group"
      >
        <div
          className="col-xs-5"
        >
          <div
            className="sc-gZMcBi iBnJdD"
          >
            <div
              className="react-autosuggest__container"
            >
              <input
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-expanded={false}
                aria-owns="react-autowhatever-1"
                autoComplete="off"
                className="react-autosuggest__input form-control"
                name="[object Object].name"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Name"
                role="combobox"
                type="text"
                value=""
              />
              <div
                className="react-autosuggest__suggestions-container"
                id="react-autowhatever-1"
              />
            </div>
          </div>
        </div>
        <div
          className="col-xs-5"
        >
          <input
            className="form-control"
            id="header.[object Object]"
            name="[object Object].value"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Value"
            type="text"
            value=""
          />
        </div>
        <div
          className="col-xs-2"
        >
          <button
            className="sc-dnqmqq gktkBI sc-bxivhb cnFuVL"
            id="removeHeaderButton0"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="sc-ifAKCX fQgDUr"
            >
              <div
                className="sc-EHOje ddddKN"
              />
              <span
                className="sc-bZQynM kagPRr"
              >
                Remove header
              </span>
            </div>
            <i
              className="fas fa-trash"
              role="presentation"
            />
          </button>
        </div>
      </div>
      <div
        className="form-group"
      >
        <div
          className="col-xs-5"
        >
          <div
            className="sc-gZMcBi iBnJdD"
          >
            <div
              className="react-autosuggest__container"
            >
              <input
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-expanded={false}
                aria-owns="react-autowhatever-1"
                autoComplete="off"
                className="react-autosuggest__input form-control"
                name="[object Object].name"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Name"
                role="combobox"
                type="text"
                value=""
              />
              <div
                className="react-autosuggest__suggestions-container"
                id="react-autowhatever-1"
              />
            </div>
          </div>
        </div>
        <div
          className="col-xs-5"
        >
          <input
            className="form-control"
            id="header.[object Object]"
            name="[object Object].value"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Value"
            type="text"
            value=""
          />
        </div>
        <div
          className="col-xs-2"
        >
          <button
            className="sc-dnqmqq gktkBI sc-bxivhb cnFuVL"
            id="removeHeaderButton1"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="sc-ifAKCX fQgDUr"
            >
              <div
                className="sc-EHOje ddddKN"
              />
              <span
                className="sc-bZQynM kagPRr"
              >
                Remove header
              </span>
            </div>
            <i
              className="fas fa-trash"
              role="presentation"
            />
          </button>
        </div>
      </div>
      <div
        className="form-group"
      >
        <div
          className="col-xs-5"
        >
          <div
            className="sc-gZMcBi iBnJdD"
          >
            <div
              className="react-autosuggest__container"
            >
              <input
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-expanded={false}
                aria-owns="react-autowhatever-1"
                autoComplete="off"
                className="react-autosuggest__input form-control"
                name="[object Object].name"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Name"
                role="combobox"
                type="text"
                value=""
              />
              <div
                className="react-autosuggest__suggestions-container"
                id="react-autowhatever-1"
              />
            </div>
          </div>
        </div>
        <div
          className="col-xs-5"
        >
          <input
            className="form-control"
            id="header.[object Object]"
            name="[object Object].value"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Value"
            type="text"
            value=""
          />
        </div>
        <div
          className="col-xs-2"
        >
          <button
            className="sc-dnqmqq gktkBI sc-bxivhb cnFuVL"
            id="removeHeaderButton2"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="sc-ifAKCX fQgDUr"
            >
              <div
                className="sc-EHOje ddddKN"
              />
              <span
                className="sc-bZQynM kagPRr"
              >
                Remove header
              </span>
            </div>
            <i
              className="fas fa-trash"
              role="presentation"
            />
          </button>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-xs-12"
        >
          <button
            className="sc-htoDjs lagtJi btn btn-link"
            disabled={false}
            id="addHeaderButton"
            onClick={[Function]}
            type="button"
          >
            <i
              className="fas fa-plus"
              role="presentation"
            />
            Add header
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
